File: src/main/java/com/github/labrynthmc/mazegen/Grid.java
package com.github.labrynthmc.mazegen;

import com.github.labrynthmc.Labrynth;
import com.sun.jna.platform.unix.X11;
import org.apache.logging.log4j.Level;

import javax.swing.plaf.basic.BasicInternalFrameTitlePane;
import java.io.Serializable;
import java.util.*;
import java.util.concurrent.atomic.AtomicInteger;

import static com.github.labrynthmc.Labrynth.LOGGER;
import static com.github.labrynthmc.Labrynth.MAZE_SIZES;

public class Grid implements Serializable {
	public static final Coords MOVE[] = {
		new Coords(0, -1), //north
		new Coords(1, 0),  //east
		new Coords(0, 1),  //south
		new Coords(-1, 0) //west
	};

	private HashMap<Coords, Cell> grid = new HashMap<>();
	private List<Coords> solution = new ArrayList<>();
	private Set<Coords> solutionSet = new HashSet<>();
	private Coords center;
	private Coords fixedCells[];
	private Coords entrance;
	private int Dx[];
	private int Dy[];
	private int size;

//	private Grid(long seed) {
//		r = new Random(seed);
//	}

	public Coords getCenter() {
		return center;
	}

	public Coords getEntrance() {
		return entrance;
	}

	public Cell getCell(Coords pos) {
		return grid.get(pos);
	}

	public Set<Coords> getKeys() {
		return grid.keySet();
	}

	public List<Coords> getSolution() {
		return solution;
	}

	public int getMinX() {
		return Dx[0];
	}
	public int getMaxX() {
		return Dx[1];
	}
	public int getMinY() {
		return Dy[0];
	}
	public int getMaxY() {
		return Dy[1];
	}

	public int getSize() {
		return size;
	}
	public void setSize(int size){ this.size = size;}

	public boolean isInSolution(Coords c) {
		return solutionSet.contains(c);
	}


	public void addCell(Coords pos) {
		if (this.grid.isEmpty()) {
			Dx = new int[]{pos.getX(), pos.getX()};
			Dy = new int[]{pos.getY(), pos.getY()};
			center = pos;
		}

		grid.put(pos, new Cell());

		if (pos.getX() < Dx[0]) Dx[0] = pos.getX();
		else if (pos.getX() > Dx[1]) Dx[1] = pos.getX();
		if (pos.getY() < Dy[0]) Dy[0] = pos.getY();
		else if (pos.getY() > Dy[1]) Dy[1] = pos.getY();
	}

	public void addPath(ArrayList<Coords> path){
		if (this.grid.get(path.get(0)) == null) this.addCell(path.get(0));
		for (int n = 1; n < path.size(); n++) {
			Coords last = path.get(n - 1);
			Coords curr = path.get(n);

			if (this.getCell(curr) == null) this.addCell(curr);
			if (curr.getX() - last.getX() == 1) {
				this.getCell(last).setSide(1, true);
				this.getCell(curr).setSide(3, true);
			} else if (curr.getX() - last.getX() == -1) {
				this.getCell(last).setSide(3, true);
				this.getCell(curr).setSide(1, true);
			} else if (curr.getY() - last.getY() == 1) {
				this.getCell(last).setSide(2, true);
				this.getCell(curr).setSide(0, true);
			} else if (curr.getY() - last.getY() == -1) {
				this.getCell(last).setSide(0, true);
				this.getCell(curr).setSide(2, true);
			}
			if (this.getCell(curr).getType() == '0')
				Labrynth.LOGGER.log(Level.ERROR,
						"Something went wrong in path with element " + n + ":"
								+ "\n\tCurrent: " + this.getCell(curr) + " " + curr
								+ "\n\tLast: " + this.getCell(last) + " " + last
				);
		}
	}

	private Coords findStart(int rad, double std, double scalar, int maxAttempts, Random r) {
		Coords start = new Coords();
		Coords center = this.getCenter();
		int attempts = 0;
		double randRad;
		double randAngle;
		do {
			randRad = r.nextGaussian() * std + rad;
			randAngle = r.nextDouble()*2*Math.PI;
			int X = (int) Math.round(randRad * Math.cos(randAngle)) + center.getX();
			int Y = (int) Math.round(randRad * Math.sin(randAngle)) + center.getY();
			start.setX(X);
			start.setY(Y);
			if (++attempts == maxAttempts) {
				attempts = 0;
				rad = (int)(scalar * Math.sqrt(this.grid.size()/Math.PI));
			}
		} while (this.getCell(start) != null);


		return start;
	}

	private ArrayList<Coords> buildPath(Coords start, Random r){
		if (this.getCell(start) != null) {
//			LOGGER.info("Start location already in maze.");
			return null;
		}
		ArrayList<Coords> path = new ArrayList<>();

		path.add(start);
		Coords pos = new Coords(start.getX(), start.getY());
		int d = r.nextInt(4);
		int rot;
		while (this.getCell(pos) == null) {
			rot = r.nextInt(3) - 1;
			d = (d + rot + 4) % 4;

			pos = pos.add(MOVE[d]);

			byte check = 0;
			for (Coords p : path) if (pos.equals(p)) {
//				LOGGER.info("Definitely hit myself.");
				check |= 1;
			}
			if (check == 0) {
				for (Coords p : fixedCells)
					if (pos.equals(p)) {
//						LOGGER.info("Definitely hit fixed cell.");
						check |= 1;
					}
			}
			if (check == 1) return null;
			path.add(pos);
		}
		return path;
	}

	//CREATE METHOD TO GENERATE STRUCTURES IN GRID

	public void genMaze(long worldSeed, int maxPaths) {
		LOGGER.info("Starting maze generation.");
		this.grid = new HashMap<>();
		this.solution = new ArrayList<>();
		this.solutionSet = new HashSet<>();

		Random r = new Random(worldSeed);

		this.setSize(maxPaths);

		this.center = new Coords(
				(int) Math.round(r.nextGaussian() * 100),
				(int) Math.round(r.nextGaussian() * 100)
		);
		Coords pos = this.center;


		this.addCell(pos);
		this.addCell(pos.add(MOVE[0]));
		this.addCell(pos.add(MOVE[1]));
		this.addCell(pos.add(MOVE[2]));
		this.addCell(pos.add(MOVE[3]));

		this.getCell(pos).setOpenSides("1111");
		this.getCell(pos.add(MOVE[0])).setOpenSides("1111");
		this.getCell(pos.add(MOVE[1])).setOpenSides("1111");
		this.getCell(pos.add(MOVE[2])).setOpenSides("1111");
		this.getCell(pos.add(MOVE[3])).setOpenSides("1111");

		this.addCell(pos.add(MOVE[0]).add(MOVE[3]));
		this.addCell(pos.add(MOVE[0]).add(MOVE[1]));
		this.addCell(pos.add(MOVE[2]).add(MOVE[3]));
		this.addCell(pos.add(MOVE[2]).add(MOVE[1]));
		this.getCell(pos.add(MOVE[0]).add(MOVE[3])).setOpenSides("0110");
		this.getCell(pos.add(MOVE[0]).add(MOVE[1])).setOpenSides("0011");
		this.getCell(pos.add(MOVE[2]).add(MOVE[3])).setOpenSides("1100");
		this.getCell(pos.add(MOVE[2]).add(MOVE[1])).setOpenSides("1001");

		this.addCell(pos.add(MOVE[0]).add(MOVE[0]));
		this.addCell(pos.add(MOVE[1]).add(MOVE[1]));
		this.addCell(pos.add(MOVE[2]).add(MOVE[2]));
		this.addCell(pos.add(MOVE[3]).add(MOVE[3]));
		this.getCell(pos.add(MOVE[0]).add(MOVE[0])).setOpenSides("0010");
		this.getCell(pos.add(MOVE[1]).add(MOVE[1])).setOpenSides("0001");
		this.getCell(pos.add(MOVE[2]).add(MOVE[2])).setOpenSides("1000");
		this.getCell(pos.add(MOVE[3]).add(MOVE[3])).setOpenSides("0100");

		this.fixedCells = new Coords[]{
				pos.add(MOVE[0]).add(MOVE[1]),
				pos.add(MOVE[0]).add(MOVE[3]),
				pos.add(MOVE[2]).add(MOVE[1]),
				pos.add(MOVE[2]).add(MOVE[3])
		};

		for (int d=0; d < 4; d++) {
			ArrayList<Coords> path = new ArrayList<>();
			pos = center.add(MOVE[d]).add(MOVE[d]);
			path.add(pos);
			for (int n = 0; n < 4; n++)
			{
				pos = pos.add(MOVE[d]);
				path.add(pos);
			}
			this.addPath(path);
		}

		double std = 2;
		double scalar = 2;
		double rad = scalar*Math.sqrt(this.grid.size()/Math.PI);
		int maxAttempts = 100;
		int lz = 4;
		while (lz < 0.25*maxPaths) {
			boolean pathFound = false;
			ArrayList<Coords> path = new ArrayList<>();
			while (!pathFound){
				int attempts = 0;
				while (attempts < maxAttempts){
					double radius = r.nextGaussian()*std + rad;
					double theta = r.nextDouble()*2*Math.PI;
					Coords start = new Coords(
							(int) (radius*Math.cos(theta))+center.getX(),
							(int) (radius*Math.sin(theta))+center.getY()
					);

					path = this.buildPath(start,r);

					if (path != null){
						LOGGER.info("Round 1:  Path found. Attempt = " + attempts + " PATH # = " + lz);
						pathFound = true;
						break;
					}
					else {
//						LOGGER.info("PATH NOT FOUND! Attempt = " + attempts + " PATH # = " + lz);
						attempts++;
					}
				}
				if (pathFound) break;

				rad = scalar*Math.sqrt(this.grid.size()/Math.PI);
				LOGGER.info(""+lz+" "+rad);
			}
			this.addPath(path);
			lz++;
//			LOGGER.info("Path " + lz + " added.");
		}
		rad = 0;
		int maxDist = Math.max(Math.abs(this.center.getY() - this.Dy[0]) ,Math.abs(this.center.getY() - this.Dy[1]));
		std = 2*maxDist;
		while (lz < maxPaths) {
			boolean pathFound = false;
			ArrayList<Coords> path = new ArrayList<>();
			while (!pathFound){
				int attempts = 0;
				while (attempts < maxAttempts){
					double radius = r.nextGaussian()*std + rad;
					double theta = r.nextDouble()*2*Math.PI;
					Coords start = new Coords(
							(int) (radius*Math.cos(theta))+center.getX(),
							(int) (radius*Math.sin(theta))+center.getY()
					);

					path = this.buildPath(start,r);

					if (path != null){
						LOGGER.info("Round 2: Path found. Attempt = " + attempts + " PATH # = " + lz);
						pathFound = true;
						break;
					}
					else {
//						LOGGER.info("PATH NOT FOUND! Attempt = " + attempts + " PATH # = " + (int)(lz+0.25*maxPaths));
						attempts++;
					}
				}
				if (pathFound) break;

//				rad += 2*std;
//				LOGGER.info(""+lz+0.25*max+" "+rad);
			}
			this.addPath(path);
			lz++;
//			LOGGER.info("Path " + (int)(lz+0.25*maxPaths) + " added.");
		}

//		-8029957180441823171
		this.createEntrance(r);
	}

	private void createEntrance(Random r) {
		Set<Coords> visited = new HashSet<>();
		Queue<Coords> queue = new LinkedList<>();
		HashMap<Coords, Coords> nextCoordToCenter = new HashMap<>();
		queue.add(this.getCenter());

		Coords lastCandidate = null;
		while (!queue.isEmpty()) {
			Coords coords = queue.poll();
			if (visited.contains(coords)) {
				continue;
			}
			visited.add(coords);
			boolean isOnSide = getCell(coords.add(MOVE[0])) == null || getCell(coords.add(MOVE[1])) == null
					|| getCell(coords.add(MOVE[2])) == null || getCell(coords.add(MOVE[3])) == null;
			if (isOnSide) {
				lastCandidate = coords;
			}
			List<Integer> sides = Arrays.asList(0,1,2,3);
			Collections.shuffle(sides, r); // Add new sides in a random order
			for (int side : sides) {
				Cell c = getCell(coords);
				if (c.getOpenSides()[side] == 1) {
					Coords next = coords.add(MOVE[side]);
					if (!visited.contains(next)) {
						queue.add(next);
						nextCoordToCenter.put(next, coords);
					}
				}
			}

		}

		if (lastCandidate != null) {
			entrance = lastCandidate;
		} else {
			entrance = center;
		}

//		List<Integer> sides = Arrays.asList(0,1,2,3);
//		Collections.shuffle(sides, r); // Add new sides in a random order
//		for (int side : sides) {
//			Cell c = getCell(entrance);
//			if (getCell(entrance.add(MOVE[side])) == null) {
//				c.setOpenSide(side, true);
//				break;
//			}
//		}

		Coords c = entrance;
		while(!c.equals(center)) {
			solution.add(c);
			c = nextCoordToCenter.get(c);
		}
		solution.add(c);
		solutionSet.addAll(solution);
	}

}

File: src/test/python/Grid.py
import numpy as np
import matplotlib.pyplot as plt
from Cell import Cell

class Grid:
	__MOVE = [
		np.array([-1,0]),
		np.array([0,1]),
		np.array([1,0]),
		np.array([0,-1]),
	]
	center = tuple()
	maze = dict()
	size = int()

	__maxDist = int()
	__FIXED = []

	def __addCell(self,pos):


		if (self.maze.get(tuple(pos)) == None):
			pos = tuple(pos)
			self.maze[pos] = Cell()
		else:
			return

		if (len(self.maze) == 1):
			self.__maxDist = 1
		elif (int(np.linalg.norm(pos - self.center)) > self.__maxDist):
			self.__maxDist = int(np.linalg.norm(pos - self.center))

	def __addPath(self,path):
		self.__addCell(path[0])
		for m in range(1,len(path)):
			last = tuple(path[m-1])
			curr = tuple(path[m])

			if (self.maze.get(curr) == None):
				self.__addCell(curr)

			disp = np.array(last) - np.array(curr)
			if (disp[0] == 1):
				self.maze[last].setSide(0,True)
				self.maze[curr].setSide(2,True)
			elif (disp[0] == -1):
				self.maze[last].setSide(2,True)
				self.maze[curr].setSide(0,True)
			elif (disp[1] == 1):
				self.maze[last].setSide(3,True)
				self.maze[curr].setSide(1,True)
			elif (disp[1] == -1):
				self.maze[last].setSide(1,True)
				self.maze[curr].setSide(3,True)

	def __buildPath(self,start,startDir):
		if (tuple(start) in self.maze):
			return None
		move = startDir

		path = [list(start)]
		# print(path)
		curr = start
		while (tuple(curr) not in self.maze):
			rot = int(np.random.uniform(0,3))-1
			move = (rot+move)%len(self.__MOVE)
			next = curr+self.__MOVE[move]

			if (list(next) in path or list(next) in self.__FIXED):
				return None

			curr = next
			path = [list(curr)]+path

		return path



	def showMaze(self):
		# print(self.__maxX)
		# print(self.__minX)
		# print(self.__maxY)
		# print(self.__minY)

		rad = self.__maxDist
		width = 2*rad
		height = 2*rad

		# print("Width = " + str(width))
		# print("Height = " + str(height))
		imCenter = np.array([rad,rad])

		im = np.zeros((5*width+5,5*height+5))
		# print(im.shape)
		for pos in self.maze:
			pos = np.array(pos)
			ind = tuple(pos-self.center+imCenter)

			m = ind[0]%(width+1)
			n = ind[1]%(height+1)
			# print("Im array: (pos " + str(tuple(pos-self.center)) + ") (inds " + str(ind) + ")")
			# print(im[5*m:5*(m+1),5*n:5*(n+1)].shape)
			# print("Cell array")
			# print(self.maze[tuple(pos)].getArray().shape)
			# print()
			im[5*m:5*(m+1),5*n:5*(n+1)] += self.maze[tuple(pos)].getArray()

		plt.imsave("maze_"+str(self.size)+".png",-im,cmap='gray')
		# plt.matshow(-im,cmap='gray')
		# plt.show()


	def genMaze(self,seed,maxPaths):
		self.size = maxPaths
		np.random.seed(seed)

		self.center = np.random.multivariate_normal([0,0],[[100,0],[0,100]])
		self.center = self.center.astype(int)
		self.center = np.array([0,0])
		# print(self.center)

		self.__addCell(self.center)
		self.__addCell(self.center+self.__MOVE[0])
		self.__addCell(self.center+self.__MOVE[1])
		self.__addCell(self.center+self.__MOVE[2])
		self.__addCell(self.center+self.__MOVE[3])
		self.maze[tuple(self.center)].setOpenSides("1111")
		self.maze[tuple(self.center+self.__MOVE[0])].setOpenSides("1111")
		self.maze[tuple(self.center+self.__MOVE[1])].setOpenSides("1111")
		self.maze[tuple(self.center+self.__MOVE[2])].setOpenSides("1111")
		self.maze[tuple(self.center+self.__MOVE[3])].setOpenSides("1111")

		self.__addCell(self.center+self.__MOVE[0]+self.__MOVE[3])
		self.__addCell(self.center+self.__MOVE[0]+self.__MOVE[1])
		self.__addCell(self.center+self.__MOVE[2]+self.__MOVE[3])
		self.__addCell(self.center+self.__MOVE[2]+self.__MOVE[1])
		self.maze[tuple(self.center+self.__MOVE[0]+self.__MOVE[3])].setOpenSides("0110")
		self.maze[tuple(self.center+self.__MOVE[0]+self.__MOVE[1])].setOpenSides("0011")
		self.maze[tuple(self.center+self.__MOVE[2]+self.__MOVE[3])].setOpenSides("1100")
		self.maze[tuple(self.center+self.__MOVE[2]+self.__MOVE[1])].setOpenSides("1001")

		self.__addCell(self.center+2*self.__MOVE[0])
		self.__addCell(self.center+2*self.__MOVE[1])
		self.__addCell(self.center+2*self.__MOVE[2])
		self.__addCell(self.center+2*self.__MOVE[3])
		self.maze[tuple(self.center+2*self.__MOVE[0])].setOpenSides("0010")
		self.maze[tuple(self.center+2*self.__MOVE[1])].setOpenSides("0001")
		self.maze[tuple(self.center+2*self.__MOVE[2])].setOpenSides("1000")
		self.maze[tuple(self.center+2*self.__MOVE[3])].setOpenSides("0100")

		self.__FIXED = [
			list(self.center+self.__MOVE[0]+self.__MOVE[3]),
			list(self.center+self.__MOVE[0]+self.__MOVE[1]),
			list(self.center+self.__MOVE[2]+self.__MOVE[3]),
			list(self.center+self.__MOVE[2]+self.__MOVE[1]),
		]

		for m in range(4):
			move = m
			start = self.center + 2*self.__MOVE[m]
			# print(str(m) + ": Starting")
			path = [list(start)]
			# print(path)
			curr = start
			while (len(path) < 5):
				rot = int(np.random.uniform(0,3))-1
				rot = 0
				move = (rot+move)%len(self.__MOVE)
				next = curr+self.__MOVE[move]

				if (list(next) in path or self.maze.get(tuple(next)) != None):
					path = [list(start)]
					move = m
					curr = start
					continue

				curr = next
				path = [list(curr)]+path

			self.__addPath(path)
			# print(str(m)+": Ending")


		scalar = 2
		r = scalar*np.sqrt(len(self.maze)/np.pi)
		# r = 0
		print("r = " + str(r))
		std = 2
		countFailed = 0
		for m in range(maxPaths-4):
			pathFound = False
			while (not pathFound):
				trials = 0
				while (trials < 100):
					R = np.random.normal(r,std)
					theta = np.random.uniform(0,2*np.pi)
					start = np.array([
							int(R*np.cos(theta)),
							int(R*np.sin(theta))
						])
					path = self.__buildPath(start,int(np.random.uniform(0,4)))
					if (path != None):
						pathFound = True
						break
					else:
						trials += 1
				# print(path)
				if (pathFound):
					# print(m/maxPaths)
					break

				r = scalar*np.sqrt(len(self.maze)/np.pi)
				# r += std
				print("Failed: r = " + str(r) + "\t size = " + str(m+4))
				countFailed += 1

			self.__addPath(path)
		# print("Failures = " + str(countFailed))
